### PR TITLE
[eclipse/xtext#1569] Refactor Xtend to Java

### PR DESCRIPTION
**Pull request by kthoms**
Replaced usage of Xtend in project org.eclipse.xtext.util, except for
active annotations.

Performed refactorings see
https://github.com/eclipse/xtext/wiki/DevGuide_RefactorXtend2Java

Signed-off-by: Karsten Thoms <karsten.thoms@itemis.de>

### CODE REVIEW

File: org.eclipse.xtext.util/src/org/eclipse/xtext/util/IFileSystemScanner.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+				if (files != null) {
+					for (File f : files) {
+						scanRec(f, acceptor);
+					}
+				}
+			}
+		}
+	}
+
+	void scan(final URI root, final IAcceptor<URI> acceptor);
```
Comment (cdietrich) [inline org.eclipse.xtext.util/src/org/eclipse/xtext/util/IFileSystemScanner.java]: final

Comment (kthoms) [reply org.eclipse.xtext.util/src/org/eclipse/xtext/util/IFileSystemScanner.java]: done

File: org.eclipse.xtext.util/src/org/eclipse/xtext/util/IFileSystemScanner.xtend [hunk 2]
```diff
@@ -1,53 +0,0 @@
...
-    	
-        @Inject extension UriExtensions
-        
-        override scan(URI root, IAcceptor<URI> acceptor) {
-            val file = new File(root.toFileString)
-            scanRec(file, acceptor)
-        }
-        
-        def void scanRec(File file, IAcceptor<URI> acceptor) {
-            // we need to convert the given file to a decoded emf file uri
```
Comment (szarnekow) [inline org.eclipse.xtext.util/src/org/eclipse/xtext/util/IFileSystemScanner.xtend]: Comment got lost

File: org.eclipse.xtext.util/src/org/eclipse/xtext/util/LineAndColumn.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+		return line;
+	}
+
+	public int getColumn() {
+		return column;
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(line, column);
```
Comment (szarnekow) [inline org.eclipse.xtext.util/src/org/eclipse/xtext/util/LineAndColumn.java]: Allocates an array on each invocation -I’m not a big fan

File: org.eclipse.xtext.util/src/org/eclipse/xtext/util/XtextVersion.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+			}
+			return version;
+		} catch (IOException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(version);
```
Comment (szarnekow) [inline org.eclipse.xtext.util/src/org/eclipse/xtext/util/XtextVersion.java]: Same

File: org.eclipse.xtext.util/src/org/eclipse/xtext/util/LineAndColumn.java (hunk 3 above)
Comment (szarnekow) [reply org.eclipse.xtext.util/src/org/eclipse/xtext/util/LineAndColumn.java]: Also: boxing if I’m not mistaken

File: org.eclipse.xtext.util/src/org/eclipse/xtext/util/IDisposable.java [hunk 5]
```diff
@@ -12,5 +12,5 @@
  */
 @SuppressWarnings("all")
```
Comment (tivervac) [inline org.eclipse.xtext.util/src/org/eclipse/xtext/util/IDisposable.java]: Missed this one

File: org.eclipse.xtext.util/src/org/eclipse/xtext/util/UriExtensions.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+import org.eclipse.emf.common.util.URI;
+
+/**
+ * @author Christian Dietrich - Initial contribution and API
+ * @since 2.14
+ */
+public class UriExtensions {
+	/**
+	 * returns a URI with empty authority, if absent and has file scheme.
```
Comment (tivervac) [inline org.eclipse.xtext.util/src/org/eclipse/xtext/util/UriExtensions.java]: This comment doesn't make any sense, and doesn't start with a capital letter.

`if absent and has file scheme` if what's absent? If what has a file scheme?
Same comment for the other method comments in this class

File: org.eclipse.xtext.util/src/org/eclipse/xtext/util/XtextVersion.java [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+import org.eclipse.emf.common.EMFPlugin;
+import org.eclipse.emf.common.util.ResourceLocator;
+import org.eclipse.xtext.xbase.lib.Exceptions;
+import org.eclipse.xtext.xbase.lib.Pure;
+
+/**
+ * @since 2.9
+ */
+@SuppressWarnings("all")
```
Comment (tivervac) [inline org.eclipse.xtext.util/src/org/eclipse/xtext/util/XtextVersion.java]: Missed one

File: org.eclipse.xtext.util/src/org/eclipse/xtext/util/internal/AlternateJdkLoader.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	private final ConcurrentMap<String, Object> locks = new ConcurrentHashMap<>();
+
+	public AlternateJdkLoader(Iterable<File> files) {
+		super(mapToURLs(files));
+	}
+	
+	private static URL[] mapToURLs (Iterable<File> files) {
+		ArrayList<File> fileList = Lists.newArrayList(files);
+		URL[] result = new URL[fileList.size()];
+		for (int i=0; i<fileList.size(); i++) {
```
Comment (tivervac) [inline org.eclipse.xtext.util/src/org/eclipse/xtext/util/internal/AlternateJdkLoader.java]: Formatting missed this

File: org.eclipse.xtext.util/src/org/eclipse/xtext/util/internal/AlternateJdkLoader.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+	public URL getResource(String name) {
+		return findResource(name);
+	}
+
+	@Override
+	public Enumeration<URL> getResources(String name) throws IOException {
+		return findResources(name);
+	}
+
+	private Object getClassLoadingLockJdk5(String className) {
```
Comment (tivervac) [inline org.eclipse.xtext.util/src/org/eclipse/xtext/util/internal/AlternateJdkLoader.java]: General question, is JDK 5 still supported? If not, shouldn't we get rid of this?

File: org.eclipse.xtext.util/src/org/eclipse/xtext/util/LineAndColumn.java (hunk 3 above)
Comment (kthoms) [reply org.eclipse.xtext.util/src/org/eclipse/xtext/util/LineAndColumn.java]: I see, thanks for this tip.

File: org.eclipse.xtext.util/src/org/eclipse/xtext/util/internal/AlternateJdkLoader.java (hunk 9 above)
Comment (kthoms) [reply org.eclipse.xtext.util/src/org/eclipse/xtext/util/internal/AlternateJdkLoader.java]: We only support Java 8. Getting rid off such things would be desirable.

File: org.eclipse.xtext.util/src/org/eclipse/xtext/util/UriExtensions.java (hunk 6 above)
Comment (kthoms) [reply org.eclipse.xtext.util/src/org/eclipse/xtext/util/UriExtensions.java]: I have added a test case and made the comments hopefully a bit more explicit.

File: org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/server/UriExtensionsTest.xtend [hunk 10]
```diff
@@ -276,6 +276,20 @@ class UriExtensionsTest {
 		var uri = createFileURI(directory.absolutePath + "/").toUriString.toUri
 		assertTrue(uri.isPrefix)
 	}
+	
+	@Test
+	def void test_withEmptyAuthority () {
+		val uriWithoutAuthority = org.eclipse.emf.common.util.URI.createURI("file:/path/to/resource/")
```
Comment (tivervac) [inline org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/server/UriExtensionsTest.xtend]: Any reason you don't just write "URI" (without the FQN) here?

File: org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/server/UriExtensionsTest.xtend [hunk 11]
```diff
@@ -276,6 +276,20 @@ class UriExtensionsTest {
...
+		assertNull(uriWithoutAuthority.authority)
+		val uriWithEmptyAuthority = uriWithoutAuthority.withEmptyAuthority
+		assertNotSame(uriWithoutAuthority, uriWithEmptyAuthority)
+		assertEquals("Returned URI is expected to have an empty instead of null authority", "", uriWithEmptyAuthority.authority)
+		
+		val uriWithAuthority = org.eclipse.emf.common.util.URI.createURI("https://xtext.org/path/to/resource/")
```
Comment (tivervac) [inline org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/server/UriExtensionsTest.xtend]: Same remark

File: org.eclipse.xtext.util/src/org/eclipse/xtext/util/internal/AlternateJdkLoader.java (hunk 9 above)
Comment (tivervac) [reply org.eclipse.xtext.util/src/org/eclipse/xtext/util/internal/AlternateJdkLoader.java]: Can you make a ticket for it? If not, I'll do it

File: org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/server/UriExtensionsTest.xtend (hunk 10 above)
Comment (cdietrich) [reply org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/server/UriExtensionsTest.xtend]: Name clash?

Comment (kthoms) [reply org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/server/UriExtensionsTest.xtend]: Yes, net.URI is already imported

**Pull request merged**